compiler/optimizer/VPHandlers.cpp

```diff
@@ -9429,39 +9494,26 @@ static TR::Node *constrainIfcmpeqne(OMR::ValuePropagation *vp, TR::Node *node, b
       TR::Node *callNode = node->getVirtualCallNodeForGuard();
 
       if (callNode &&
-          callNode->getOpCode().isCallIndirect())
+          callNode->getOpCode().isCall())
          {
-         TR::SymbolReference *symRef        = callNode->getSymbolReference();
-         TR::MethodSymbol    *methodSymbol  = symRef->getSymbol()->castToMethodSymbol();
-
-         int32_t firstArgIndex = callNode->getFirstArgumentIndex();
-         bool isGlobal;
-         TR::VPConstraint *constraint = vp->getConstraint(callNode->getChild(firstArgIndex), isGlobal);
-         TR_OpaqueClassBlock *thisType;
-         //dumpOptDetails(vp->comp(), "Guard %p Call %p constraint %p\n", node, callNode, constraint);
-         if (constraint &&
-             constraint->isFixedClass() &&
-             (thisType = constraint->getClass()) &&
-              methodSymbol->isVirtual())
-            {
-            TR::ResolvedMethodSymbol * resolvedMethodSymbol = methodSymbol->getResolvedMethodSymbol();
-            if (resolvedMethodSymbol)
-               {
-               TR_ResolvedMethod *originalResolvedMethod = resolvedMethodSymbol->getResolvedMethod();
-               TR_OpaqueClassBlock *originalMethodClass = originalResolvedMethod->classOfMethod();
-
-               if ((vp->fe()->isInstanceOf(thisType, originalMethodClass, true, true) == TR_yes) &&
-                   !originalResolvedMethod->virtualMethodIsOverridden())
-                 {
-                 virtualGuardConstraint = constraint;
-                 TR_VirtualGuard *virtualGuard = vp->comp()->findVirtualGuardInfo(node);
-                 if (virtualGuard && virtualGuard->canBeRemoved())
-                    virtualGuardWillBeEliminated = true;
-                 }
-              }
-           }
-        }
-     }
+         virtualGuardWillBeEliminated = canFoldNonOverriddenGuard(vp, callNode, node);
+         if (!virtualGuardWillBeEliminated &&
+             vp->lastTimeThrough() &&
+             vp->_isGlobalPropagation)
+            {
+            dumpOptDetails(vp->comp(), "Saving guard %p for call %p\n", node, callNode);
+            TR_Pair<TR::TreeTop, TR::CFGEdge> *pair = new (vp->comp()->trHeapMemory())
+                                                   TR_Pair<TR::TreeTop, TR::CFGEdge>(vp->_curTree, edge);
+            List<TR_Pair<TR::TreeTop, TR::CFGEdge>> &list = (vp->_callNodeToGuardNodes)[callNode->getGlobalIndex()];
+            list.setRegion(vp->comp()->trMemory()->heapMemoryRegion());
```

Comment:
Is the stack memory region freed too early? I think this list is only needed within the same pass of GVP

Reply:
Changed.